In [50]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler

import warnings
warnings.filterwarnings('ignore')

In [51]:
#importing district wise crime data
district_crime_data=pd.read_csv(r'https://raw.githubusercontent.com/manishkumbhare/Capstone-Project/main/datasets/01_District_wise_crimes_committed_IPC_2001_2012.csv')
district_crime_data.head()

STATE/UT       DISTRICT  YEAR  MURDER  ATTEMPT TO MURDER  \
0  Andhra Pradesh       ADILABAD  2001     101                 60   
1  Andhra Pradesh      ANANTAPUR  2001     151                125   
2  Andhra Pradesh       CHITTOOR  2001     101                 57   
3  Andhra Pradesh       CUDDAPAH  2001      80                 53   
4  Andhra Pradesh  EAST GODAVARI  2001      82                 67   

   CULPABLE HOMICIDE NOT AMOUNTING TO MURDER  RAPE  CUSTODIAL RAPE  \
0                                         17    50               0   
1                                          1    23               0   
2                                          2    27               0   
3                                          1    20               0   
4                                          1    23               0   

   OTHER RAPE  KIDNAPPING & ABDUCTION  ...  ARSON  HURT/GREVIOUS HURT  \
0          50                      46  ...     30                1131   
1          23                      53  ...     69                1543   
2          27                      59  ...     38                2088   
3          20                      25  ...     23                 795   
4          23                      49  ...     41                1244   

   DOWRY DEATHS  ASSAULT ON WOMEN WITH INTENT TO OUTRAGE HER MODESTY  \
0            16                                                149     
1             7                                                118     
2            14                                                112     
3            17                                                126     
4            12                                                109     

   INSULT TO MODESTY OF WOMEN  CRUELTY BY HUSBAND OR HIS RELATIVES  \
0                          34                                  175   
1                          24                                  154   
2                          83                                  186   
3                          38                                   57   
4                          58                                  247   

   IMPORTATION OF GIRLS FROM FOREIGN COUNTRIES  CAUSING DEATH BY NEGLIGENCE  \
0                                            0                          181   
1                                            0                          270   
2                                            0                          404   
3                                            0                          233   
4                                            0                          431   

   OTHER IPC CRIMES  TOTAL IPC CRIMES  
0              1518              4154  
1               754              4125  
2              1262              5818  
3              1181              3140  
4              2313              6507  

[5 rows x 33 columns]

### # Requirement No - 4.1
#### Create 3 clusters as below.
1. Sensitive Area's
2. Moderate Area's
3. Peaceful Area's

#### As we require to do Area wise clustering, So We shall be Choosing some major crimes that defines on the area and which are not domestic crimes like - Dowry, Arson, Breach of trust etc.
#### SO we shall be choosing - Murder, Rape, Kidnapping, Robbery, Burglary, Theft, Roits only.

In [52]:
df=district_crime_data[['STATE/UT','DISTRICT', 'YEAR', 'MURDER', 'RAPE', 'KIDNAPPING & ABDUCTION', 'ROBBERY', 'BURGLARY', 'THEFT', 'RIOTS',]]

df.head()

STATE/UT       DISTRICT  YEAR  MURDER  RAPE  KIDNAPPING & ABDUCTION  \
0  Andhra Pradesh       ADILABAD  2001     101    50                      46   
1  Andhra Pradesh      ANANTAPUR  2001     151    23                      53   
2  Andhra Pradesh       CHITTOOR  2001     101    27                      59   
3  Andhra Pradesh       CUDDAPAH  2001      80    20                      25   
4  Andhra Pradesh  EAST GODAVARI  2001      82    23                      49   

   ROBBERY  BURGLARY  THEFT  RIOTS  
0       41       198    199     78  
1       16       191    366    168  
2       14       237    723    156  
3        4        98    173    164  
4       25       437   1021     70

In [53]:
# Creating a new column named 'Total_crime' with the sum of all the select crime type
df['Total_crime']=df[['MURDER','RAPE','KIDNAPPING & ABDUCTION','ROBBERY','BURGLARY','THEFT','RIOTS' ]].sum(axis=1)

df.head()

STATE/UT       DISTRICT  YEAR  MURDER  RAPE  KIDNAPPING & ABDUCTION  \
0  Andhra Pradesh       ADILABAD  2001     101    50                      46   
1  Andhra Pradesh      ANANTAPUR  2001     151    23                      53   
2  Andhra Pradesh       CHITTOOR  2001     101    27                      59   
3  Andhra Pradesh       CUDDAPAH  2001      80    20                      25   
4  Andhra Pradesh  EAST GODAVARI  2001      82    23                      49   

   ROBBERY  BURGLARY  THEFT  RIOTS  Total_crime  
0       41       198    199     78          713  
1       16       191    366    168          968  
2       14       237    723    156         1317  
3        4        98    173    164          564  
4       25       437   1021     70         1707

In [54]:
# Selecting 'Total_crime' as features for clustering
features = df[['Total_crime']]  # Add more features as needed

# Standardizing the features
scaler = StandardScaler()
features_standardized = scaler.fit_transform(features)

# Applying k-means clustering
kmeans = KMeans(n_clusters=3, random_state=42)
df['CLUSTER'] = kmeans.fit_predict(features_standardized)

# Defining cluster names
cluster_names = {
    0: 'Sensitive Area',
    1: 'Moderate Area',
    2: 'Peaceful Area',
}

# Maping cluster names to the dataframe
df['CLUSTER_NAME'] = df['CLUSTER'].map(cluster_names)

# Displaying the result
print(df[['STATE/UT', 'DISTRICT', 'YEAR', 'CLUSTER_NAME']])

# Assigning the result to a dataframe 'cluster_result'





            STATE/UT       DISTRICT  YEAR    CLUSTER_NAME
0     Andhra Pradesh       ADILABAD  2001  Sensitive Area
1     Andhra Pradesh      ANANTAPUR  2001  Sensitive Area
2     Andhra Pradesh       CHITTOOR  2001  Sensitive Area
3     Andhra Pradesh       CUDDAPAH  2001  Sensitive Area
4     Andhra Pradesh  EAST GODAVARI  2001  Sensitive Area
...              ...            ...   ...             ...
8592           Delhi            STF  2012  Sensitive Area
8593           Delhi           WEST  2012   Moderate Area
8594     Lakshadweep    LAKSHADWEEP  2012  Sensitive Area
8595     Pondicherry       KARAIKAL  2012  Sensitive Area
8596     Pondicherry    Pondicherry  2012  Sensitive Area

[8597 rows x 4 columns]


### # Requirement No - 4.2 : 
#### Create DataFrame for each cluster that shows data according to the areas.

In [55]:
cluster_result=df[['STATE/UT', 'DISTRICT', 'YEAR', 'CLUSTER_NAME']]

In [56]:
df['CLUSTER_NAME'].value_counts()

Sensitive Area    8045
Moderate Area      527
Peaceful Area       25
Name: CLUSTER_NAME, dtype: int64

#### So majorly we have have 'Sensitive Areas' compared to 'Moderate Areas' and 'Peaceful Areas'

In [57]:
# Creating a new DataFrame for "Sensitive Area"
sensitive_area_df = df[df['CLUSTER_NAME'] == 'Sensitive Area'][['STATE/UT', 'DISTRICT', 'YEAR', 'Total_crime', 'CLUSTER_NAME']]

# Displaying the new DataFrame for "Sensitive Area"
print("DataFrame for Sensitive Area:")
sensitive_area_df.sample(5)

DataFrame for Sensitive Area:


STATE/UT            DISTRICT  YEAR  Total_crime    CLUSTER_NAME
4054  Uttar Pradesh           GHAZIABAD  2006         1621  Sensitive Area
8241    Maharashtra           OSMANABAD  2012          868  Sensitive Area
7663     Tamil Nadu  THIRUNELVELI RURAL  2011          830  Sensitive Area
3217         Odisha       SRP(ROURKELA)  2005          122  Sensitive Area
2984      Jharkhand             DEOGHAR  2005          571  Sensitive Area

In [58]:
# Creating a new DataFrame for "Moderate Area"
moderate_area_df = df[df['CLUSTER_NAME'] == 'Moderate Area'][['STATE/UT', 'DISTRICT', 'YEAR', 'Total_crime', 'CLUSTER_NAME']]

# Displaying the new DataFrame for "Moderate Area"
print("DataFrame for Moderate Area:")
moderate_area_df.sample(5)

DataFrame for Moderate Area:


STATE/UT      DISTRICT  YEAR  Total_crime   CLUSTER_NAME
2053          Delhi         SOUTH  2003         4061  Moderate Area
6970  Uttar Pradesh  KANPUR NAGAR  2010         2213  Moderate Area
6697    Maharashtra    AHMEDNAGAR  2010         3058  Moderate Area
7815          Delhi    SOUTH-WEST  2011         1965  Moderate Area
7487    Maharashtra  THANE COMMR.  2011         4937  Moderate Area

In [60]:
# Creating a new DataFrame for "Peaceful Area"
peaceful_area_df = df[df['CLUSTER_NAME'] == 'Peaceful Area'][['STATE/UT', 'DISTRICT', 'YEAR', 'Total_crime', 'CLUSTER_NAME']]

# Displaying the new DataFrame for "Peaceful Area"
print("DataFrame for Peaceful Area:")
peaceful_area_df.sample(5)

DataFrame for Peaceful Area:


STATE/UT        DISTRICT  YEAR  Total_crime   CLUSTER_NAME
7465     Maharashtra   MUMBAI COMMR.  2011        17219  Peaceful Area
5975     Maharashtra   MUMBAI COMMR.  2009        15947  Peaceful Area
356      Maharashtra          MUMBAI  2001        15599  Peaceful Area
2425     Maharashtra          MUMBAI  2004        15621  Peaceful Area
1373  Andhra Pradesh  HYDERABAD CITY  2003         9107  Peaceful Area

### # Requirement No - 4.3 : 
#### Analyse your clusters and prepare a report that explains all your observations.

### #Report

#### Overview:
India has been categorized into three clusters based on crime data analysis, each reflecting different levels of criminal activities.  
    
#### Clusters:  
Sensitive Area:
  
Total Incidents: 8045  
Description: This cluster represents areas where crime rates are relatively high, indicating a need for heightened attention and law enforcement strategies.  
'SITAMARHI' from Bihar, 'DIMAPUR' from Nagaland, 'RATLAM' from Madhya Pradesh, 'KOLLAM RURAL' form Kerala, 'VELLORE' from Timal Nadu coming under this cluster.

    
Moderate Area:  
  
Total Incidents: 527  
Description: Areas falling into this cluster experience a moderate level of criminal activities. Focused efforts on crime prevention and community engagement may be beneficial.  
Districts like 'GUNTUR', 'OUTER Delhi', 'GWALIOR', 'JAIPUR EAST', 'RAIPUR', 'PUNE COMMR.' are comming under this cluster.
        
Peaceful Area:  

Total Incidents: 25  
Description: This cluster denotes regions with low crime rates, signifying a generally peaceful environment. However, continuous monitoring and proactive measures should be in place to maintain this status.  
Mostly 'MUMBAI COMMR.', 'Bangalore COMMR.' and 'Hydrabad City' is comming under this category so we can say these areas are Safest in India.

#### Recommendations:
Sensitive Area:

Implement targeted law enforcement initiatives to address specific crime trends.  
Enhance community policing and public awareness programs.  
Collaborate with local communities to identify and address root causes of crime.    
  
  
Moderate Area:

Strengthen community engagement and local law enforcement partnerships.  
Implement crime prevention programs tailored to the specific challenges faced in these areas.  
Focus on education and awareness to empower communities in preventing and reporting crimes.    
  
  
Peaceful Area:  

Maintain vigilance to prevent any potential rise in criminal activities.  
Invest in community programs that promote safety and well-being.  
Share best practices with other areas to sustain the peaceful environment.  


#### Conclusion:  
This crime report provides valuable insights into the varying levels of criminal activities across different regions in India. It emphasizes the importance of tailored strategies to address the unique challenges faced by each cluster. By fostering collaboration between law enforcement agencies, local communities, and policymakers, India can work towards creating safer and more secure environments for its residents.

### # Requirement No 4.4 :   
#### Capstone project overall story in your own words. Min 1000 words.

#### this requirement is being satisfied in the Project Overview sheet.